blueflood-core/src/main/java/com/rackspacecloud/blueflood/types/Token.java

```diff
@@ -0,0 +1,147 @@
+package com.rackspacecloud.blueflood.types;
+
+import org.apache.commons.lang.StringUtils;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+
+import static java.util.stream.Collectors.joining;
+import static java.util.stream.Collectors.toList;
+import static java.util.stream.Collectors.toSet;
+
+/**
+ *
+ * For a given tenantId = '111111', metricName = 'a.b.c.d',
+ *
+ * level 2 token will have the following information.
+ *
+ *      token = 'c'
+ *      parent = 'a.b'
+ *      isLeaf = false
+ *      id = '111111:a.b.c'
+ *      tenantId = '111111'
+ *
+ * level 3 token which is a leaf node will have the following information
+ *
+ *      token = 'd'
+ *      parent = 'a.b.c'
+ *      isLeaf = true
+ *      id = '111111:a.b.c.d:$'
+ *      tenantId = '111111'
+ *
+ */
+public class Token {
+
+    public static final String LEAF_NODE_SUFFIX = ":$";
+    public static final String SEPARATOR = ":";
```

Comment:
We use this separator today to define a locator which is of the form <tenantid>:<metric_name>. This will be a problem for very specific scenarios where `:$` is the end of metric token name. For example: If a locator name happens to be `111:foo.bar.baz:$.qux`, we create tokens with id's `[111:foo, 111:foo.bar, **111:foo.bar.baz:$**, 111:foo.bar.baz:$.qux:$]` No problems yet. But if there is another metric which is `111:foo.bar.baz`,  we create tokens is id's `[111:foo, 111:foo.bar, **111:foo.bar.baz:$**]`. 

For metric `111:foo.bar.baz:$.qux`, **111:foo.bar.baz:$** is not a leaf node, but for metric `111:foo.bar.baz`, **111:foo.bar.baz:$** is a leaf node.

I think we need to add documentation that for metric names, metric tokens should not end with `:$`